### ISSUE TITLE
CSS @import prettify removes content

### ISSUE THREAD

**Issue by sodik82** [prettier/prettier#2584]
I am not sure if this is an issue - but when I am prettify  CSS (see [playground](https://prettier.io/playground/#%7B%22content%22%3A%22%40import%20%5C%22..%2Fassets%2Ffont%2Fleague-gothic%2Fleague-gothic.css%5C%22%5Cn%2F*%20css%20*%2F%5Cn%5Cn.ManagerPage%20%7B%5Cn%20%20height%3A%20100%25%3B%5Cn%7D%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22es5%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22postcss%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%2C%22ast%22%3Afalse%7D%7D))
```css
@import "../assets/font/league-gothic/league-gothic.css"
/* css */

.ManagerPage {
  height: 100%;
}
```
it removes some content that should be kept there (in playground it ends up as `@import .ManagerPage;`)

The problem is solved when I "properly" use semi-colon, e.g.
```css
@import "../assets/font/league-gothic/league-gothic.css";

.ManagerPage {
  height: 100%;
}
```

The best solution would be if prettier could add semicolon for the import. Or at least not to discard my content if it can't fix it.

**Comment by vjeux** [comment]
Looks like an issue with our regex that tries to figure out if we're using less or scss (doing the wrong thing).

**Comment by lydell** [comment]
A couple of notes here:

- The example is invalid CSS (according to http://www.css-validator.org/#validate_by_input), invalid SCSS (doesn't compile, https://www.sassmeister.com/) and invalid Less (doesn't compile, http://less2css.org/), but PostCSS parses it just fine (as usual).

- The SCSS parser happens to do just fine in this case, while the Less parser (which ends up being used) leaves out lots of the code from the AST.

- There is nothing wrong with the "likely SCSS check" in this case.

- postcss-less and postcss-scss parse this example in different ways:

```
> console.log(util.inspect(require("postcss-less").parse('@import a b'), {colors: true, depth: Infinity}))
Root {
  raws: { semicolon: false, after: '' },
  type: 'root',
  nodes: 
   [ Import {
       raws: { before: '', afterName: ' ', after: ' ' },
       type: 'import',
       nodes: [],
       name: 'import',
       parent: [Circular],
       source: 
        { start: { line: 1, column: 1 },
          input: Input { css: '@import a b', id: '<input css 2>' },
          end: { line: 0, column: 0 } },
       importPath: 'b',
       directives: '' } ],
  source: 
   { input: Input { css: '@import a b', id: '<input css 2>' },
     start: { line: 1, column: 1 } } }
undefined
> console.log(util.inspect(require("postcss-scss").parse('@import a b'), {colors: true, depth: Infinity}))
Root {
  raws: { semicolon: false, after: '' },
  type: 'root',
  nodes: 
   [ AtRule {
       raws: { before: '', between: '', afterName: ' ' },
       type: 'atrule',
       name: 'import',
       parent: [Circular],
       source: 
        { start: { line: 1, column: 1 },
          input: Input { css: '@import a b', id: '<input css 1>' },
          end: { line: 1, column: 11 } },
       params: 'a b' } ],
  source: 
   { input: Input { css: '@import a b', id: '<input css 1>' },
     start: { line: 1, column: 1 } } }
```

Conclusion: There's definitely a bug in the postcss-less parser that parses this. It shouldn't ignore chunks of the input. But what it actually _should_ do is a bit fuzzy, since this is invalid Less in the first place. I guess doing something similar to the postcss-scss parser would be fine.

**Comment by lydell** [comment]
I now understand the problem better.

There's one major issue, and one smaller "issue" to think about.

### Background

This is the test case:

```css
@import "../assets/font/league-gothic/league-gothic.css"

.ManagerPage {
  height: 100%;
}
```

Both postcss-less and postcss-scss parse the above as if it was written like this:

```css
@import "../assets/font/league-gothic/league-gothic.css" .ManagerPage {
  height: 100%;
}
```

If I hack src/parser-postcss.js to _always_ use postcss-scss, the above is also the output from prettier.

My guess is that they parse it like this to support multiline media queries.

### The major issue

postcss-less only keeps the last "word" of the at-rule stuff before the braces. That's why there's missing code in Prettier's output. This needs to be fixed in the postcss-less parser. When that's done, both postcss-less and postcss-scss seem to have the exact same behavior.

### The smaller "issue" to think about

As mentioned in the previous comment, the test case is invalid CSS, invalid SCSS and invalid Less, but postcss, postcss-scss and postcss-less all parse it.

To me, if Prettier formats the test case into this:

```css
@import "../assets/font/league-gothic/league-gothic.css" .ManagerPage {
  height: 100%;
}
```

... that would be totally fine. The user could split it up again, and add the missing semicolon.

But there are also other ways to think about this:

- Since the test case is invalid code, perhaps postcss, postcss-scss and postcss-less should throw an error for this. But perhaps they are not very interested in such a breaking change.
- postcss, postcss-scss and postcss-less could change the way the consume tokens for the `@import` at-rule (and perhaps others?), so that the test case parses as an `@import` rule followed by a regular rule with a selector.

I'm not sure what the best solution is here.

### Conclusion

1. Fix the "major" postcss-less bug. Then at least no code is lost anymore.
2. Possibly do something more, or not. Needs discussion.

**Comment by lydell** [comment]
If we trust our tests, this patch fixes the problem and breaks no tests:

```diff
--- a/src/parser-postcss.js
+++ b/src/parser-postcss.js
@@ -209,7 +209,7 @@ function requireParser(isSCSS) {
 }

 function parse(text /*, parsers, opts*/) {
-  const isLikelySCSS = !!text.match(/(\w\s*: [^}:]+|#){|@import[^\n]+(url|,)/);
+  const isLikelySCSS = true;
   try {
     return parseWithParser(requireParser(isLikelySCSS), text);
   } catch (e) {
```

**Comment by lydell** [comment]
https://github.com/shellscape/postcss-less/issues/88

**Comment by vjeux** [comment]
"If we trust our tests"

I wouldn't trust our tests when it comes to CSS, we haven't done any real attempts at breaking it with running it through big codebases (I did it for pure css, but not less/sass) nor fuzzing.

But this line is super brittle and I wonder if we could do better.

**Comment by lydell** [comment]
@vjeux Totally agree with that.

**Comment by azz** [comment]
Should we close this one and open another, or remove High-Pri?

**Comment by lydell** [comment]
Isn't this issue still both relevant and high-pri?

**Comment by azz** [comment]
Yes you're right. I hadn't read through #2646 yet. Was just going by the title.

**Comment by lydell** [comment]
The postcss-less issue is not super easy to fix, so I found a workaround until then: #2652

**Comment by vjeux** [comment]
One thing we've done in the past is for rules that we couldn't easily re-print, is to just take the range and print it as it was in the source. I'm not sure if the range is correct in this case, but it may be a solution.
